Handle DaisyUI variants and nested layer rules

nicky9door opened the pull request:
This PR adds the unocss equivalent of the tailwind addVariant function.  This should rectify part of issue #27.

I'm not much of a Typescript guy so feel free to let me know if there are any syntax changes you would like made

nicky9door committed:
Implement add variant handler

nicky9door committed:
Refactor flattenRules to handle mixed declarations and nested rules/atrules

nicky9door commented:
I pushed a second commit which refactors the flattenRules to allow traversal of 'mixed nodes' (i.e. nodes which contain declarations and nested rules/atrules)

This refactor should correctly handle the recent DaisyUI changes which introduced css layer rules in most components.  I tried to do things the 'UnoCSS' way by passing an object that uses the UnoCSS symbols for parent selector and css layer.  Note that the UnoCSS config will need to enable the ```outputToCssLayers``` config option for actual CSS layers to be used, otherwise the default UnoCSS strategy of simply ordering rules without the actual ```@layer``` rules being present.  Also note that this means that the generated CSS file will likely have 'scattered' rules (e.g ```.btn``` main rules will be in the ```daisy``` layer whereas ```.btn-primary``` will appear later in the ```daisyui.modifier.color``` layer)

I did a quick test with daisyui@5.2.5 which was the version before css layers were introduced.  The CSS generated without apparent issue but more testing might be needed.  I'm also not sure if this will work on v4 but I can't think of anything specific that would cause breakage.

Neraoll commented on src/index.ts:
```diff
@@ -23,32 +23,48 @@ interface Options {
+    if (node.type == 'atrule') {
       if (node.name === 'keyframes') {
-        yield node.toString();
+        return node.toString();
```
Using `return` here seems to create a bug with the radio checked animation in the demo, "keyframes" at-rules are not being parsed/generated, probably because the `return` stop the loop on `nodes`.

I was able to fix it by going back to a `yield` and adding a `continue;` after this line to match the old behavior.

Neraoll replied:
Using `return` here seems to create a bug with the radio checked animation in the demo, `@keyframes` rules are not being parsed/generated, probably because the `return` stop the loop on `nodes`.

I was able to fix it by going back to a `yield` and adding a `continue;` after this line to match the old behavior.

```suggestion
        yield node.toString();
        continue;
```

Neraoll commented on src/index.ts:
```diff
@@ -23,32 +23,48 @@ interface Options {
   NOMERGE = /^file-input(?:-.+)?|.*::-webkit-slider-runnable-track$/;
 
-function* flattenRules(nodes: ChildNode[], parents: string[] = []): Generator<[string[], string, Declaration[]] | string> {
+function* flattenRules(nodes: ChildNode[], parentUnoSymbols: Object = {}): Generator<[string[], string, Declaration[]] | string> {
```
`Object` is not usually used in TypeScript as it's a standard built-in type that `"Provides functionality common to all JavaScript objects."` like constructor, toString(), ... (more on that [here](https://medium.com/@zong8498/exploring-the-differences-between-object-and-object-in-typescript-927b50c9e0de) or [here](https://stackoverflow.com/a/49465172)).

So we should use either `object`(lowercase) or declare a more specific type like :
```
type ParentUnoSymbols = Partial<{
  [symbols.layer]: string
  [symbols.parent]: string
  [symbols.selector]: string
}>;
```

And to avoid TypeScript errors we should also update the function return type to match the new behavior as it no longer yield `[string[], string, Declaration[]] | string` but `[ParentUnoSymbols, Declaration[]] | string`.

Neraoll replied:
`Object` is not usually used in TypeScript as it's a standard built-in type that `"Provides functionality common to all JavaScript objects."` like constructor, toString(), ... (more on that [here](https://medium.com/@zong8498/exploring-the-differences-between-object-and-object-in-typescript-927b50c9e0de) or [here](https://stackoverflow.com/a/49465172)).

So we should use either `object`(lowercase) or declare a more specific type like :
```
type ParentUnoSymbols = Partial<{
  [symbols.layer]: string
  [symbols.parent]: string
  [symbols.selector]: string
}>;
```

And to avoid TypeScript errors we should also update the function return type to match the new logic as it no longer yield `[string[], string, Declaration[]] | string` but `[ParentUnoSymbols, Declaration[]] | string`.

```suggestion
type ParentUnoSymbols = Partial<{
  [symbols.layer]: string
  [symbols.parent]: string
  [symbols.selector]: string
}>;

function* flattenRules(nodes: ChildNode[], parentUnoSymbols: ParentUnoSymbols = {}): Generator<[ParentUnoSymbols, Declaration[]] | string> {
```

Neraoll commented on src/index.ts:
```diff
@@ -131,26 +153,42 @@ export async function presetDaisy(options?: Options): Promise<Preset<Record<stri
       );
     },
+
+    addVariant(name, selector) {
```
To improve/fix arguments type checking we can use a sub-type of the `addVariant` function in tailwind like :

```suggestion
    addVariant(name: string, selector: string | string[]) {
```

Neraoll commented on src/index.ts:
```diff
@@ -23,32 +23,48 @@ interface Options {
-        continue;
-      }
+    let unoSymbols = {...parentUnoSymbols}
+    let selector = node.selector;
```
TypeScript is not happy with `node.selector` because the property is not available in all type of `ChildNode`.

I suggest implementing a match of the old early return/continue logic to [narrow the type](https://www.typescriptlang.org/docs/handbook/2/narrowing.html) of `node` :
```suggestion
    if (node.type !== 'atrule' && node.type !== 'rule') {
      if (node.type !== 'comment') {
        // eslint-disable-next-line no-console
        console.warn('skipping', node.type);
      }

      continue;
    }

    if (!node.nodes || node.nodes.length === 0) {
      continue;
    }

    let unoSymbols = {...parentUnoSymbols}
    let selector = ('selector' in node ? node.selector : undefined);
```

Neraoll commented on src/index.ts:
```diff
@@ -131,26 +153,42 @@ export async function presetDaisy(options?: Options): Promise<Preset<Record<stri
+
+          return {
+            matcher: matcher.slice(name.length + 1),
+            selector: s => `${s}${selector}`
```
With the new sub-type from the tailwind PluginAPI `selector` can now be an array so we need to handle that : 
```suggestion
            selector: (s) => `${s}${(Array.isArray(selector) ? selector.join('') : selector)}`
```

Neraoll reviewed (commented):
Great work !! 😃

Except for the little bug from the change on [line 33](https://github.com/ameinhardt/unocss-preset-daisy/pull/28/files#diff-a2a171449d862fe29692ce031981047d7ab755ae7f84c707aef80701b3ea0c80R33) all the other comments I got are to improve/fix errors from TypeScript type-checking (via `pnpm tsc --noEmit`).

I will also suggest to run `pnpm lint` and fix the issues (or I can provide review/code suggestions for them too).

A part form that everything seems to work on the demo page with the latest version of daisyui to date v5.5.14.

We may just have to change (again) the layers used for `addBase`, `addComponents` and `addUtilities` because since [v5.5.1](https://daisyui.com/docs/changelog/#551-2025-11-13) it look like they have renamed/restructure them : https://github.com/saadeghi/daisyui/commit/c20d12c92c51cefb5f26c95a9449ae3e9c3c4b11


I also tested the changes on the demo page with the UnoCSS config option `outputToCssLayers: true,` to allow for real `@layer` usage in the generated CSS.
But enabling this make the use/setup of the reset `@unocss/reset/tailwind.css` a bit more complicated.
As of my knowledge, the only way to make it work properly is to create a new .css file like `reset.css` containing this :
```
@import '@unocss/reset/tailwind.css' layer(base);
```
To wrap the reset in the tailwind `base` layer, otherwise all reset rules overrides all rules wrapped in `@layer` by tailwind/daisyui.

And then import the file in the `index.html` before `uno.css` :
```
<script type="module">
	import './reset.css'
	import 'uno.css'
</script>
```

nicky9door committed:
Update to latest daisy ui

nicky9door committed:
Add additional types for parent symbols and node types to improve type safety

nicky9door commented:
@Neraoll Thanks for the input.  Sorry for the delay but I finally got around to implementing your suggestions (plus a little more typing + null safety suggested by pnpm lint).  I did a quick test and it seems to work with the demo page

Regarding the reset, I believe the new presetWind4 includes a reset that get its own css layer. I did a quick update where I switched the demo to use wind4 and removed the explicit import of the reset css in the html file and it seemed to work

sevmonster commented:
In order to build your fork in-tree under Bun, I had to do the following:
```sh
#!/bin/sh
d=node_modules/@ameinhardt/unocss-preset-daisy
sed -Ei 's#("types": "\./dist/index\.d\.)m(ts")#\1\2#g' $d/package.json
sed -Ei "s#(entries: \\[)'\\./src/index'(\\],)#\\1{builder:'mkdist',input:'./src/',outDir:'./dist/'}\\2#g" $d/build.config.js
bun --cwd=node_modules/@ameinhardt/unocss-preset-daisy i
sed -Ei 's#(import parse from "\./parser\.)(js";)#\1m\2#g' $d/dist/index.mjs
```

The discovery process went like this:
1. I don't know if this is a Rollup or Unbuild bug, but since the filesystem path has the name from `package.json` name in it (`@ameinhardt/unocss-preset-daisy`), Rollup thinks `src/index`/`src/parser` are external, and refuses to bundle them.
2. I first tried to fix this by forcing Rollup to not mark the paths as external by putting this in `build.config.js`, which worked:
    ```js
      hooks: {
        'rollup:options'(ctx, options) {
          const f = options.external;
          options.external = (id, importer, isResolved) =>
            id.includes('@ameinhardt/unocss-preset-daisy')
              ? false
              : originalExternal(id, importer, isResolved);
        }
      }
    ```
3. ...however, I then got this error:
    ```
    [2:21:20 PM]  ERROR  Error building node_modules/@ameinhardt/unocss-preset-daisy: RollupError: Expected ',', got '{' in node_modules/@ameinhardt/unocss-preset-daisy/src/index.ts
    ```
    I don't know if that is a Bun issue or what, but I decided it wasn't worth my time to continue debugging...
4. Next, I forced Unbuild to not use Rollup by replacing the entrypoints:
    ```js
    entries: [
      {
        builder: 'mkdist',
        input: './src/',
        outDir: './dist/'
      }
    ]
    ```
5. Possibly as a consequence, sone of the paths don't exist (`index.d.mts`, `parser.js`), so I fixed those.

After all that, I was able to get usable `dist` files and import the module.

Don't know how much of this is relevant, and I don't have much idea what I'm doing.
